refactor(compiler-cli): make the output AST translator generic

petebacondarwin opened the pull request:
This commit refactors the `ExpressionTranslatorVisitor` so that it
is not tied directly to the TypeScript AST. Instead it uses generic
`TExpression` and `TStatement` types that are then converted
to concrete types by the `TypeScriptAstFactory`.

This paves the way for a `BabelAstFactory` that can be used to
generate Babel AST nodes instead of TypeScript, which will be
part of the new linker tool.

petebacondarwin committed:
refactor(compiler-cli): make the output AST translator generic

This commit refactors the `ExpressionTranslatorVisitor` so that it
is not tied directly to the TypeScript AST. Instead it uses generic
`TExpression` and `TStatement` types that are then converted
to concrete types by the `TypeScriptAstFactory`.

This paves the way for a `BabelAstFactory` that can be used to
generate Babel AST nodes instead of TypeScript, which will be
part of the new linker tool.

JoostK commented on packages/compiler-cli/ngcc/src/rendering/esm_rendering_formatter.ts:
```diff
@@ -10,8 +10,9 @@ import MagicString from 'magic-string';
-import {NOOP_DEFAULT_IMPORT_RECORDER, Reexport} from '../../../src/ngtsc/imports';
+import {Reexport} from '../../../src/ngtsc/imports';
 import {Import, ImportManager, translateStatement} from '../../../src/ngtsc/translator';
+import {TypeScriptAstFactory} from '../../../src/ngtsc/translator/src/typescript_ast_factory';
```
```suggestion
import {TypeScriptAstFactory} from '../../../src/ngtsc/translator';
```

JoostK commented on packages/compiler-cli/src/ngtsc/translator/src/api/ast_factory.ts:
```diff
@@ -0,0 +1,261 @@
+   * @param range a mapping of the generated expression back to the original source, or null if
+   *     there is no mapping.
+   */
+  createArrayLiteral(elements: TExpression[], range: SourceMapRange|null): TExpression;
```
I'm a bit confused why this (and some others) takes `range`, if we also have `setSourceMapRange`

petebacondarwin replied:
It is frustrating, but here is the problem:

The translator does not have enough information to be responsible for directing the `AstFactory` to construct each piece of a template literal in an abstract way. So the `AstFactory` must be responsible for putting the final AST of a template literal together.

The translator should be in charge of assigning source-spans, as you say, by calling `setSourceMapRange()` but we need each piece of the template literal to get its own source map ranges.

So this was the compromise that I came up with. In the case of template literals we delegate attaching the source map ranges of the pieces to the factory.

...

I considered moving all the source-map range management into the factory and dropping the `setSourceMapRange()` from the `AstFactory` but equally there are places where only the translator knows that a range needs to be attached.

JoostK replied:
I still don't get why this compromise is needed, though. Anyone who's calling into e.g. `createArrayLiteral` providing a `SourceMapRange`, could also have called `factory.setSourceMapRange` on the return value for that call, couldn't they?

```ts
const node = this.factory.createArrayLiteral(elements, range);
// becomes
const node = this.factory.createArrayLiteral(elements);
this.factory.setSourceMapRange(node, range);
``` 

Am I missing something here?

petebacondarwin replied:
Doh! Sorry.
I was thinking about `TemplateLiteral` nodes not `ArrayLiteral` nodes.
Yes, you are quite right :-)

petebacondarwin replied:
Oh so actually, I was on the right track but just backwards.
Yes, we can do what you suggest but it seemed less error prone to put the calls to `setSourceMapRange()` inside the factories rather than hoping the translator would do it correctly.
But I don't think it makes much difference and it is more consistent to have a single way of doing things.

JoostK replied:
Yes, agreed; either pass `range` as last param into all factory methods and not have `setSourceMapRange`, or just have `setSourceMapRange`. I went with the latter in the spike impl as it felt cumbersome to let all factory methods have that param and need to do something with it, and it mirrored TS's API more. I don't feel too strongly about it though, so I'm fine switching to the parameter-approach.

petebacondarwin replied:
So now that I am awake, I see more clearly.

At some point I decided that passing the range to the factory was cleaner, saving the translator having to do extra range existence checks and an extra API on the `AstFactory`. Then I realised that the localize string stuff could not work this way, so had to keep the `setSourceMapRange()` method. This meant that we were stuck in the hybrid scenario in which this PR finds itself.

So moving back to the `setSourceMapRange()` call approach is the most consistent.

JoostK commented on packages/compiler-cli/src/ngtsc/transform/src/transform.ts:
```diff
@@ -360,3 +365,12 @@ function maybeFilterDecorator(
   return decorator.import !== null && decorator.import.from === '@angular/core';
 }
+
+function recorderFnFactory(defaultImportRecorder: DefaultImportRecorder): (expr: ts.Expression) =>
```
nit: I prefer

```suggestion
function createRecorderFn(defaultImportRecorder: DefaultImportRecorder): (expr: ts.Expression) =>
```

petebacondarwin replied:
As you wish

JoostK commented on packages/compiler-cli/src/ngtsc/translator/src/api/ast_factory.ts:
```diff
@@ -0,0 +1,261 @@
+  setSourceMapRange(node: TStatement|TExpression, sourceMapRange: SourceMapRange): void;
+}
+
+export type UnaryOperators = '+'|'-'|'!';
```
I feel this should be singular `UnaryOperator`/`BinaryOperator` instead of plural, as that makes it more logical in places where it's used.

petebacondarwin replied:
Yeah, I was thinking that too...

JoostK commented on packages/compiler-cli/src/ngtsc/translator/src/api/ast_factory.ts:
```diff
@@ -0,0 +1,261 @@
+  end: SourceMapLocation;
+}
+
+export interface PropertyAssignment<TExpression> {
```
These various interfaces/types may benefit from some comments.

JoostK commented on packages/compiler-cli/src/ngtsc/translator/src/api/import_generator.ts:
```diff
@@ -0,0 +1,36 @@
+
+/**
+ * The symbol name and import namespace of an imported symbol,
+ * which has been registered through the ImportManager.
```
```suggestion
 * which has been registered through the ImportGenerator.
```

JoostK commented on packages/compiler-cli/src/ngtsc/translator/src/api/import_generator.ts:
```diff
@@ -0,0 +1,36 @@
+}
+
+/**
+ *
```
😱

JoostK reviewed (commented):
I did a very quick initial pass over this, I plan to go over it again.

josephperrott reviewed (approved):
LGTM,

Reviewed-for: dev-infra

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/environment.ts:
```diff
@@ -213,8 +213,8 @@ export class Environment {
     // Use `translateExpression` to convert the `Expression` into a `ts.Expression`.
-    return translateExpression(
-        ngExpr, this.importManager, NOOP_DEFAULT_IMPORT_RECORDER, ts.ScriptTarget.ES2015);
+    return translateExpression<ts.Expression>(
```
Isn't this inferred?

petebacondarwin replied:
Actually the inferred types break the `AstFactory` -> `TypeScriptAstFactory` relationship. I can't work out why.
But as it turns out this particular typing was wrong. I will correct it. It should be:

```ts
translateExpression<ts.Statement, ts.Expression>(
```

JoostK commented on packages/compiler-cli/src/ngtsc/translator/src/typescript_ast_factory.ts:
```diff
@@ -0,0 +1,307 @@
+  ['!', ts.SyntaxKind.ExclamationToken],
+]);
+
+const BINARY_OPERATORS = new Map<BinaryOperators, ts.BinaryOperator>([
```
If this were an object literal we'd have completeness checking:

```ts
const BINARY_OPERATORS: Record<BinaryOperator, ts.BinaryOperator> = {
  '&&': ts.SyntaxKind.AmpersandAmpersandToken,
  '>': ts.SyntaxKind.GreaterThanToken,
  '>=': ts.SyntaxKind.GreaterThanEqualsToken,
  '&': ts.SyntaxKind.AmpersandToken,
  '/': ts.SyntaxKind.SlashToken,
  '==': ts.SyntaxKind.EqualsEqualsToken,
  '===': ts.SyntaxKind.EqualsEqualsEqualsToken,
  '<': ts.SyntaxKind.LessThanToken,
  '<=': ts.SyntaxKind.LessThanEqualsToken,
  '-': ts.SyntaxKind.MinusToken,
  '%': ts.SyntaxKind.PercentToken,
  '*': ts.SyntaxKind.AsteriskToken,
  '!=': ts.SyntaxKind.ExclamationEqualsToken,
  '!==': ts.SyntaxKind.ExclamationEqualsEqualsToken,
  '||': ts.SyntaxKind.BarBarToken,
  '+': ts.SyntaxKind.PlusToken,
};
```

petebacondarwin replied:
Nice!

JoostK commented on packages/compiler-cli/src/ngtsc/translator/src/typescript_ast_factory.ts:
```diff
@@ -0,0 +1,307 @@
+  createTaggedTemplate(
+      tag: ts.Expression, template: TemplateLiteral<ts.Expression>,
+      range: SourceMapRange|null): ts.Expression {
+    let templateLiteral: ts.TemplateLiteral;
```
We may want to throw here if `this.scriptTarget < ts.ScriptTarget.ES2015`.

petebacondarwin replied:
Hmm, yeah, I was wondering whether we should move the whole decision making of downleveling these tagged templates into function calls into the factory rather than the translator...

JoostK replied:
I was wondering the exact same thing yesterday. Initially it looked strange to have `downlevelLocalizedStrings` inside of the translator, but could kinda see why you did it that way. Maybe we should explore if doing if from the factory is possible (it has to generate an import! I think that's a problem as the import generator may not be something we'll want to expose to the factory?).

petebacondarwin replied:
Yeah it is a question of where to move the complexity...

petebacondarwin replied:
So now I've gone the other way and moved all language-level related knowledge out of the `AstFactory` and into the translator. This makes the factory implementation cleaner.

JoostK commented on packages/compiler-cli/src/ngtsc/translator/src/api/ast_factory.ts:
```diff
@@ -0,0 +1,290 @@
+
+  /**
+   * Create an assignment expression (e.g. `lhsExpr = rhsExpr`).
+   * @param target An expression that evaluates to the left side of the assignment.
```
Before @gkalpak comments about it, let me mention that this file is quite inconsistent with capitalization and dots after sentences in docblocks 😅

petebacondarwin replied:
😱

petebacondarwin replied:
@JoostK are you trying to put @gkalpak out of a job 😛

JoostK replied:
I hope not!!

petebacondarwin replied:
I made the file consistent: always start param descriptions with a lowercase and end with a dot.
I note that many files in this part of the code base are equally inconsistent.

AndrewKushnir commented:
@petebacondarwin it looks like there is a test failing in the `test` CirckeCI job. Please let me know if you want to run presubmit now or have the test fixed first (in case it's a legit failure). Thank you.

petebacondarwin commented:
@AndrewKushnir  - sorry, yes, I should have switched this to `WIP` since the last commit was just trying out an idea that @JoostK had. The error is legit, but we feel it is not the right direction anyway, so I will revert that last commit.

petebacondarwin commented:
OK, @AndrewKushnir - the CI is green and this is ready for a presubmit if you get a chance.

JoostK commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -942,7 +942,7 @@ describe('i18n support in the template compiler', () => {
     });
 
-    it('should sanitize ids and generate proper var names', () => {
+    it('should sanitize ids and generate proper let names', () => {
```
https://twitter.com/BenLesh/status/1305699317794181122

```suggestion
    it('should sanitize ids and generate proper var names', () => {
```

petebacondarwin replied:
Hmm I think actually:

```suggestion
    it('should sanitize ids and generate proper variable names', () => {
```

JoostK replied:
Works for me :-)

petebacondarwin replied:
To be fair, I did a find and replace on `"var "`, so `varying` would not have been changed 😛

JoostK reviewed (approved):
This looks great!

gkalpak commented on packages/compiler-cli/src/ngtsc/translator/src/typescript_ast_factory.ts:
```diff
@@ -0,0 +1,264 @@
+    return ts.createTaggedTemplate(tag, templateLiteral);
+  }
+
+  createThrowStatement(expression: ts.Expression): ts.Statement {
+    return ts.createThrow(expression);
+  }
```
Why not:

```suggestion
  createThrowStatement = ts.createThrow;
```

gkalpak commented on packages/compiler-cli/src/ngtsc/translator/src/typescript_ast_factory.ts:
```diff
@@ -0,0 +1,264 @@
+    return ts.createThrow(expression);
+  }
+
+  createTypeOfExpression(expression: ts.Expression): ts.Expression {
+    return ts.createTypeOf(expression);
+  }
```
Why not:

```suggestion
  createTypeOfExpression = ts.createTypeOf;
```

gkalpak commented on packages/compiler-cli/src/ngtsc/translator/src/typescript_ast_factory.ts:
```diff
@@ -0,0 +1,264 @@
+}
+
+// HACK: Use this in place of `ts.createTemplateMiddle()`.
+// Revert once https://github.com/microsoft/TypeScript/issues/35374 is fixed
```
```suggestion
// Revert once https://github.com/microsoft/TypeScript/issues/35374 is fixed.
```

<sub>Sorry, but I have a reputation to uphold 😇</sub>

gkalpak commented on packages/compiler-cli/src/ngtsc/translator/src/typescript_ast_factory.ts:
```diff
@@ -0,0 +1,264 @@
+}
+
+// HACK: Use this in place of `ts.createTemplateTail()`.
+// Revert once https://github.com/microsoft/TypeScript/issues/35374 is fixed
```
```suggestion
// Revert once https://github.com/microsoft/TypeScript/issues/35374 is fixed.
```

gkalpak commented on packages/compiler-cli/src/ngtsc/translator/test/typescript_ast_factory_spec.ts:
```diff
@@ -0,0 +1,389 @@
+
+  describe('attachComments()', () => {
+    it('should add the comments to the given statement', () => {
+      const {items: [stmt], generate} = setupStatements(`x = 10;`);
```
Why isn't tslint complaining about template literals without interpolations? 🤨 😠

petebacondarwin replied:
🤷  - but I will fix them

gkalpak commented on packages/compiler-cli/src/ngtsc/translator/test/typescript_ast_factory_spec.ts:
```diff
@@ -0,0 +1,389 @@
+    it('should create a call marked with a PURE comment if `pure` is true', () => {
+      const {items: [callee, arg1, arg2], generate} = setupExpressions(`foo`, `42`, `"moo"`);
+      const call = factory.createCallExpression(callee, [arg1, arg2], true);
+      expect(generate(call)).toEqual(['/*@__PURE__*/ foo(42, "moo")'].join('\n'));
```
Why not:

```suggestion
      expect(generate(call)).toEqual('/*@__PURE__*/ foo(42, "moo")');
```

petebacondarwin replied:
A leftover from when the generated output was on two lines.

gkalpak commented on packages/compiler-cli/src/ngtsc/translator/test/typescript_ast_factory_spec.ts:
```diff
@@ -0,0 +1,389 @@
+      expect(generate(literal)).toEqual('42');
+    });
+
+    it('should create a number literal', () => {
```
Duplicate test description (same wit the one above) 😱 😱 😱

gkalpak commented on packages/compiler-cli/src/ngtsc/translator/test/typescript_ast_factory_spec.ts:
```diff
@@ -0,0 +1,389 @@
+      expect(generate(literal)).toEqual('undefined');
+    });
+
+    it('should create a null literal', () => {
```
Why is `undefined` wrapped in backticks and `null` is not? 🤨

petebacondarwin replied:
Who knows? I blame the midnight faeries.

gkalpak replied:
Yeah, they're the worse 😄

gkalpak reviewed (approved):
This looks really nice and neat 🏆

<sub>I now realize I might have gotten a little too nit-picky. Feel free to ignore some of the comments 🙈</sub>

AndrewKushnir commented:
[Presubmit](http://test/OCL:332759389:BASE:332759381:1600649763757:847b8e1b).

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/test/typescript_ast_factory_spec.ts:
```diff
@@ -0,0 +1,389 @@
+         expect(generate(varDecl)).toEqual('const foo = 42;');
+       });
+
+    it('should create a downlevelled declaration statement node for the given variable name and initializer',
```
https://en.wiktionary.org/wiki/downlevelled

But I guess this is yet another Americanism that I must adhere to...

petebacondarwin replied:
I used "downlevelled" three times in the code recently. But I accept that there are 48 instances of "downleveled". 🤷

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/api/ast_factory.ts:
```diff
@@ -0,0 +1,294 @@
+ * The original location of the start or end of a node created by the `AstFactory`.
+ */
+export interface SourceMapLocation {
+  offset: number;
+  line: number;
+  column: number;
```
That is so weird. I am sure I wrote these comments before. They must have got blown away by some rebasing.

petebacondarwin replied:
Ahah it is in the `AstHost` defined in the Babel PR:

```ts
export interface Range {
  /** 0-based character position of the range start in the source file text. */
  startPos: number;
  /** 0-based line index of the range start in the source file text. */
  startLine: number;
  /** 0-based column position of the range start in the source file text. */
  startCol: number;
  /** 0-based character position of the range end in the source file text. */
  endPos: number;
}
```

I'll add some more comments here.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/import_manager.ts:
```diff
@@ -0,0 +1,52 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+import * as ts from 'typescript';
+import {ImportRewriter, NoopImportRewriter} from '../../imports/src/core';
```
Absolutely! The preference is always to import from the `index.ts` and not deeply from the `src` folder. Unfortunately, VS Code automatic imports don't follow that rule, so we need eagle-eyed reviewers to spot the mistakes.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/type_translator.ts:
```diff
@@ -0,0 +1,243 @@
+  }
+
+  visitTypeofExpr(ast: o.TypeofExpr, context: Context): ts.TypeQueryNode {
+    const typeNode = this.translateExpression(ast.expr, context);
```
You are spot on. This is an independent fix and should be in a separate commit... I'll try to extract it.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/translator.ts:
```diff
@@ -5,285 +5,266 @@
-    return context.isStatement ? result : ts.createParen(result);
+  visitWriteVarExpr(expr: o.WriteVarExpr, context: Context): TExpression {
+    const assignment = this.factory.createAssignment(
+        this.setSourceMapRange(this.factory.createIdentifier(expr.name), expr.sourceSpan),
```
I don't know if this was a bug, as such.

Source-mapping is so confusing, since the tests we have require a number of assumptions to be made about the source-maps. In general it is better to have more source-spans than less, since you get better resolution in the source-maps and source-map flattening is more accurate.

My general rule of thumb is that any identifier should have a source-span associated with it.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/translator.ts:
```diff
@@ -5,285 +5,266 @@
+    //
+    // Each message part may be prefixed with "metadata", which is wrapped in colons (:) delimiters.
+    // The metadata is attached to the first and subsequent message parts by calls to
+    // `serializeI18nHead()` and `serializeI18nTemplatePart()` respectively.
```
Actually I will reorganized these two methods since they share some logic. Then the comment should only be needed in one place.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/api/ast_factory.ts:
```diff
@@ -0,0 +1,294 @@
+
+  /**
+   * Create a ternary expression (e.g. `testExpr ? trueExpr : falseExpr`).
+   * @param condition an expression that will be tested for truthiness
+   * @param thenExpression an expression that is executed if `condition` is truthy
+   * @param elseExpression an expression that is executed if `condition` is falsy
```
🙀

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/api/ast_factory.ts:
```diff
@@ -0,0 +1,294 @@
+
+  /**
+   * Create an expression that is instantiating the `expression` as a class.
+   * @param expression an expression that evaluates to a class to be instantiated.
```
I've just gone with `constructor` rather than `class`.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/api/ast_factory.ts:
```diff
@@ -0,0 +1,294 @@
+   * @param expression an expression that evaluates to the object to be accessed.
+   * @param propertyName the name of the property to access.
+   */
+  createPropertyAccess(expression: TExpression, propertyName: string): TExpression;
```
But then you would have to wrap teh `propertyName` string into an identifier each time.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/api/ast_factory.ts:
```diff
@@ -0,0 +1,294 @@
+
+  /**
+   * Create an expression that declares a new variable, possibly initialized to `initializer`.
+   *
```
I agree - I'll add the lines.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/api/ast_factory.ts:
```diff
@@ -0,0 +1,294 @@
+/**
+ * Used to create transpiler specific AST nodes from Angular Output AST nodes in an abstract way.
+ *
+ * Note that the AstFactory makes no assumptions about the target language being generated.
```
Actually this is the only place it wasn't wrapped! Good spot, thanks!

petebacondarwin commented on packages/compiler-cli/src/ngtsc/translator/src/typescript_ast_factory.ts:
```diff
@@ -0,0 +1,264 @@
+    return ts.createParen(expression);
+  }
+
+  createPropertyAccess(expression: ts.Expression, propertyName: string): ts.Expression {
+    return ts.createPropertyAccess(expression, propertyName);
+  }
```
Good question! I had it like that at one point but the TS compiler complained... I think. But now it is not complaining locally. So, I will switch it back and see what CircleCI has to say.

petebacondarwin committed:
refactor(compiler-cli): remove undesirable cast in the type translator

The cast to `ts.Identifier` was a hack that "just happened to work".
The new approach is more robust and doesn't have to undermine
the type checker.

petebacondarwin committed:
refactor(compiler-cli): remove unused imports

These imports are not used and so are just bloating the code unnecessarily

petebacondarwin committed:
refactor(compiler): use a named type for cooked/raw string objects

Using an interface makes the code cleaner and more readable.
This change also adds the `range` property to the type to be used
for source-mapping.

petebacondarwin committed:
fix(compiler-cli): generate `let` statements in ES2015+ mode

When the target of the compiler is ES2015 or newer then we should
be generating `let` and `const` variable declarations rather than `var`.

petebacondarwin committed:
refactor(compiler-cli): split up translator file

This file contains a number of classes making it long and hard to work with.
This commit splits the `ImportManager`, `Context` and `TypeTranslatorVisitor`
classes, along with associated functions and types into their own files.

petebacondarwin committed:
refactor(compiler-cli): simplify imports from compiler to type translator

Previously each identifier was being imported individually, which made for a
very long import statement, but also obscurred, in the code, which identifiers
came from the compiler.

petebacondarwin committed:
fixup! refactor(compiler-cli): make the output AST translator generic

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:332759389:BASE:332880681:1600709534850:8a4eee2a).

alfaproject commented:
OOC, do we need to care about ES5 at all? I thought that we started delegating the responsibility of downleveling to the building tool, like Angular CLI or whatever else, right?

JoostK commented:
@alfaproject ngcc uses this machinery to emit code into ES5 bundles of libraries.

petebacondarwin commented:
In the long run we would not but for now there are plenty of libraries and project setups that rely upon ES5

alfaproject commented:
Got it. Thank you for explaining. (:
